src/adapter/src/coord/caught_up.rs

```diff
@@ -385,4 +419,139 @@ impl Coordinator {
+        let min_timestamp = now.saturating_sub(lookback_window);
+        let min_timestamp_dt = mz_ore::now::to_datetime(min_timestamp);
+
+        // Get the replica status collection GlobalId
+        let replica_status_item_id = self
+            .catalog()
+            .resolve_builtin_storage_collection(&MZ_CLUSTER_REPLICA_STATUS_HISTORY);
+        let replica_status_gid = self
+            .catalog()
+            .get_entry(&replica_status_item_id)
+            .latest_global_id();
+
+        // Acquire a read hold to determine the as_of timestamp for snapshot_and_stream
+        let read_holds = self
+            .controller
+            .storage_collections
+            .acquire_read_holds(vec![replica_status_gid])
+            .expect("can't acquire read hold for mz_cluster_replica_status_history");
+        let read_hold = if let Some(read_hold) = read_holds.into_iter().next() {
+            read_hold
+        } else {
+            // Collection is not readable anymore, but we return an empty set
+            // instead of panicing.
+            return BTreeSet::new();
+        };
+
+        let as_of = read_hold
+            .since()
+            .iter()
+            .next()
+            .cloned()
+            .expect("since should not be empty");
+
+        let mut replica_statuses_stream = self
+            .controller
+            .storage_collections
+            .snapshot_and_stream(replica_status_gid, as_of)
+            .await
+            .expect("can't read mz_cluster_replica_status_history");
+
+        let mut replica_problem_counts: BTreeMap<ReplicaId, u32> = BTreeMap::new();
+
+        while let Some((source_data, _ts, diff)) = replica_statuses_stream.next().await {
+            // Only process inserts (positive diffs)
+            if diff <= 0 {
+                continue;
+            }
+
+            // Extract the Row from SourceData
+            let row = match source_data.0 {
+                Ok(row) => row,
+                Err(_) => continue, // Skip rows with errors
+            };
+
+            let mut iter = row.into_iter();
+
+            let replica_id: ReplicaId = iter
+                .next()
+                .expect("missing replica_id")
+                .unwrap_str()
+                .parse()
+                .expect("must parse as replica ID");
+            let _process_id = iter.next().expect("missing process_id").unwrap_uint64();
+            let status = iter
+                .next()
+                .expect("missing status")
+                .unwrap_str()
+                .to_string();
+            let reason_datum = iter.next().expect("missing reason");
+            let reason = if reason_datum.is_null() {
+                None
+            } else {
+                Some(reason_datum.unwrap_str().to_string())
+            };
+            let occurred_at = iter
+                .next()
+                .expect("missing occurred_at")
+                .unwrap_timestamptz();
```

Comment:
This is in `mz_internal`, so the schema might change. You could add a comment at the definition of `MZ_CLUSTER_REPLICA_STATUS_HISTORY` calling out that this code also needs to be changed if somebody wants to change the schema.

Reply:
I was hoping tests would highlight that, and I don't know how likely it is that the person would read this comment. IMO it's more likely they will notice when this code fails.

Reply:
Well, yes, some Nightly upgrade check would hopefully catch it.